### PR TITLE
RNN derivatives

### PR DESCRIPTION
**Pull request by schaefertim**
Introduce RNN derivatives and test them.
Additionally, there are many changes to docstrings and type hints. As a result, the modified files pass pydocstyle and darglint checks.

### CODE REVIEW

**Comment by f-dangel** [comment]
Side note: coverage drop is due to unsupported arguments in `RNNDerivatives` (see [here](https://coveralls.io/builds/40236842/source?filename=backpack%2Fcore%2Fderivatives%2Frnn.py#L36-47)).

**Comment by f-dangel** [comment]
For internal documentation: mapping between derivatives and modules (`derivatives_for`) moved from main library to tests (`test.core.derivatives`).

File: test/core/derivatives/rnn_settings.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+###############################################################################
+#                                   examples                                  #
+###############################################################################
+
+example = {
+    "module_fn": lambda: torch.nn.RNN(input_size=5, hidden_size=3),
+    "input_fn": lambda: torch.rand(size=(10, 8, 5)),
+}
+RNN_SETTINGS.append(example)
```
Comment (f-dangel) [inline test/core/derivatives/rnn_settings.py:28]: The example is very similar to the 'real' test cases. I think removing it does not complicate writing a new case -> remove?

File: backpack/core/derivatives/shape_check.py [hunk 2]
```diff
@@ -39,113 +52,143 @@ def check_shape(mat, like, diff=1):
...
+def _mat_prod_accept_vectors(
+    mat_prod: Callable[..., Tensor],
+    vec_criterion: Callable[[Tensor, Module, Any, Any], bool],
+) -> Callable[..., Tensor]:
     """Add support for vectors to matrix products.
 
     vec_criterion(mat, module) returns if mat is a vector.
+
+    Args:
+        mat_prod: .
```
Comment (f-dangel) [inline backpack/core/derivatives/shape_check.py:86]: Function that processes multiple vectors in format of a matrix.

File: backpack/core/derivatives/shape_check.py [hunk 3]
```diff
@@ -39,113 +52,143 @@ def check_shape(mat, like, diff=1):
...
+        vec_criterion: .
```
Comment (f-dangel) [inline backpack/core/derivatives/shape_check.py:87]: Function that returns if an input is a single vector that must be formatted into a matrix first before processing

File: backpack/core/derivatives/shape_check.py [hunk 4]
```diff
@@ -39,113 +52,143 @@ def check_shape(mat, like, diff=1):
...
+
+    Returns:
+        .
```
Comment (f-dangel) [inline backpack/core/derivatives/shape_check.py:90]: Wrapped `mat_prod` function that processes multiple vectors in format of a matrix and supports vector-shaped inputs which are internally converted to the correct format.

Preserves format of input:
- If the input format is a vector, the output format is a vector.
- If the input format is a matrix, the output format is a matrix.

File: backpack/core/derivatives/shape_check.py [hunk 5]
```diff
@@ -39,113 +52,143 @@ def check_shape(mat, like, diff=1):
...
-def mat_prod_check_shapes(mat_prod, in_check, out_check):
-    """Check that input and output have correct shapes."""
+def mat_prod_check_shapes(
+    mat_prod: Callable, in_check: Callable, out_check: Callable
+) -> Callable[..., Tensor]:
+    """Check that input and output have correct shapes.
+
+    Args:
+        mat_prod: .
+        in_check: .
```
Comment (f-dangel) [inline backpack/core/derivatives/shape_check.py:154]: Function that checks the input to `mat_prod`

File: backpack/core/derivatives/shape_check.py [hunk 6]
```diff
@@ -39,113 +52,143 @@ def check_shape(mat, like, diff=1):
...
+        out_check: .
```
Comment (f-dangel) [inline backpack/core/derivatives/shape_check.py:155]: Function that checks the output of `mat_prod`

File: backpack/core/derivatives/shape_check.py [hunk 7]
```diff
@@ -39,113 +52,143 @@ def check_shape(mat, like, diff=1):
...
+
+    Returns:
+        .
```
Comment (f-dangel) [inline backpack/core/derivatives/shape_check.py:158]: Wrapped `mat_prod` function with input and output checks

File: backpack/core/derivatives/shape_check.py [hunk 8]
```diff
@@ -39,113 +52,143 @@ def check_shape(mat, like, diff=1):
...
 ###############################################################################
-def mat_prod_check_shapes(mat_prod, in_check, out_check):
-    """Check that input and output have correct shapes."""
+def mat_prod_check_shapes(
+    mat_prod: Callable, in_check: Callable, out_check: Callable
+) -> Callable[..., Tensor]:
+    """Check that input and output have correct shapes.
+
+    Args:
+        mat_prod: .
```
Comment (f-dangel) [inline backpack/core/derivatives/shape_check.py:153]: Function that applies a derivative operator to multiple vectors handed in as a matrix

File: backpack/core/derivatives/shape_check.py [hunk 9]
```diff
@@ -185,44 +243,64 @@ def wrapped_mat_prod_check_shapes(self, module, g_inp, g_out, mat, *args, **kwar
...
 
 # TODO Refactor using partials
-def make_hessian_mat_prod_accept_vectors(make_hessian_mat_prod):
+def make_hessian_mat_prod_accept_vectors(
+    make_hessian_mat_prod: Callable,
+) -> Callable[..., Callable[..., Tensor]]:
+    """Accept vectors for hessian_mat_prod.
+
+    Args:
+        make_hessian_mat_prod: .
```
Comment (f-dangel) [inline backpack/core/derivatives/shape_check.py:258]: Creates multiplication routine of a matrix with the module Hessian

File: backpack/core/derivatives/shape_check.py [hunk 10]
```diff
@@ -185,44 +243,64 @@ def wrapped_mat_prod_check_shapes(self, module, g_inp, g_out, mat, *args, **kwar
...
 
-        return new_hessian_mat_prod
+        return _new_hessian_mat_prod
+
+    return _wrapped_make_hessian_mat_prod
 
-    return wrapped_make_hessian_mat_prod
 
+def make_hessian_mat_prod_check_shapes(make_hessian_mat_prod: Callable) -> Callable:
+    """Accept shapes for hessian_mat_prod.
```
Comment (f-dangel) [inline backpack/core/derivatives/shape_check.py:283]: Wrap hessian_mat_prod with shape checks for input and output

File: backpack/core/derivatives/shape_check.py [hunk 11]
```diff
@@ -185,44 +243,64 @@ def wrapped_mat_prod_check_shapes(self, module, g_inp, g_out, mat, *args, **kwar
...
+
+    Args:
+        make_hessian_mat_prod: .
```
Comment (f-dangel) [inline backpack/core/derivatives/shape_check.py:286]: Creates multiplication routine of a matrix with the module Hessian

File: backpack/core/derivatives/shape_check.py [hunk 12]
```diff
@@ -185,44 +243,64 @@ def wrapped_mat_prod_check_shapes(self, module, g_inp, g_out, mat, *args, **kwar
...
+
+    Returns:
+        .
```
Comment (f-dangel) [inline backpack/core/derivatives/shape_check.py:289]: Wrapped hessian_mat_prod with shape checks for input and output

File: backpack/core/derivatives/shape_check.py [hunk 13]
```diff
@@ -185,44 +243,64 @@ def wrapped_mat_prod_check_shapes(self, module, g_inp, g_out, mat, *args, **kwar
...
+def make_hessian_mat_prod_accept_vectors(
+    make_hessian_mat_prod: Callable,
+) -> Callable[..., Callable[..., Tensor]]:
+    """Accept vectors for hessian_mat_prod.
+
+    Args:
+        make_hessian_mat_prod: .
+
+    Returns:
+        .
```
Comment (f-dangel) [inline backpack/core/derivatives/shape_check.py:261]: Wrapped hessian_mat_prod which converts vector-format inputs to a matrix before processing. Preserves format of input.

File: test/core/derivatives/derivatives_test.py [hunk 14]
```diff
@@ -46,26 +47,42 @@
...
     problem.set_up()
     mat = torch.rand(V, *problem.input_shape).to(problem.device)
 
     backpack_res = BackpackDerivatives(problem).jac_mat_prod(mat)
-    autograd_res = AutogradDerivatives(problem).jac_mat_prod(mat)
+    if all(string in request.node.callspec.id for string in ["RNN", "cuda"]):
+        # torch does not implement cuda double-backwards pass on RNNs and
+        # recommends this workaround
+        with torch.backends.cudnn.flags(enabled=False):
+            autograd_res = AutogradDerivatives(problem).jac_mat_prod(mat)
```
Comment (f-dangel) [inline test/core/derivatives/derivatives_test.py:75]: Move to `AutogradDerivatives`

File: test/core/derivatives/derivatives_test.py [hunk 15]
```diff
@@ -46,26 +47,42 @@
...
 ]
 
+RNN_PROBLEMS = make_test_problems(RNN_SETTINGS)
+RNN_IDS = [problem.make_id() for problem in RNN_PROBLEMS]
 
-@pytest.mark.parametrize("problem", NO_LOSS_PROBLEMS, ids=NO_LOSS_IDS)
-def test_jac_mat_prod(problem, V=3):
+
+# selecting only every second rnn problem (only cpu problems)
+# because torch is not able to calculate this function on CUDA
```
Comment (f-dangel) [inline test/core/derivatives/derivatives_test.py:55]: Is the selection statement true? The code looks as if the `autograd` computation is run with a backend other than `cudnn`

File: test/core/derivatives/implementation/autograd.py [hunk 16]
```diff
@@ -10,39 +12,56 @@
 class AutogradDerivatives(DerivativesImplementation):
     """Derivative implementations with autograd."""
 
-    def jac_vec_prod(self, vec):
+    def jac_vec_prod(self, vec) -> Tensor:
+        """Jacobian vector product.
```
Comment (f-dangel) [inline test/core/derivatives/implementation/autograd.py:16]: Input-output Jacobian

File: test/core/derivatives/implementation/backpack.py [hunk 17]
```diff
@@ -1,65 +1,108 @@
...
+        self.store_forward_io()
+        return self.problem.derivative.weight_ih_l0_jac_t_mat_prod(
+            self.problem.module, None, None, mat, sum_batch=sum_batch
+        )
+
+    def weight_hh_l0_jac_t_mat_prod(self, mat, sum_batch):  # noqa: D102
+        self.store_forward_io()
+        return self.problem.derivative.weight_hh_l0_jac_t_mat_prod(
+            self.problem.module, None, None, mat, sum_batch=sum_batch
+        )
```
Comment (f-dangel) [inline test/core/derivatives/implementation/backpack.py:85]: Refactor the body of all `param_jac_t`s into a documented `_param_jac_t_mat_prod(self, mat, sum_batch, param_str)` to reduce duplication.

File: test/core/derivatives/implementation/backpack.py [hunk 18]
```diff
@@ -1,65 +1,108 @@
...
+    def sum_hessian(self):  # noqa: D102
         self.store_forward_io()
         return self.problem.derivative.sum_hessian(self.problem.module, None, None)
 
-    def input_hessian_via_sqrt_hessian(self, mc_samples=None):
+    def input_hessian_via_sqrt_hessian(self, mc_samples=None) -> Tensor:
+        """Computes the input hessian.
+
+        Args:
+            mc_samples: number of samples. Defaults to None.
```
Comment (f-dangel) [inline test/core/derivatives/implementation/backpack.py:101]: `None` uses the exact Hessian. `int` uses an MC approximation with the specified number of samples.

File: test/core/derivatives/implementation/backpack.py [hunk 19]
```diff
@@ -1,65 +1,108 @@
...
+
+        Returns:
+            hessian
+        """
         # MC_SAMPLES = 100000
```
Comment (f-dangel) [inline test/core/derivatives/implementation/backpack.py:106]: dead code -> remove

File: test/core/derivatives/implementation/base.py [hunk 20]
```diff
@@ -1,23 +1,163 @@
...
+
+        Args:
+            problem: test problem
+        """
         self.problem = problem
 
-    def jac_mat_prod(self, mat):
+    @abstractmethod
+    def jac_mat_prod(self, mat: Tensor) -> Tensor:
+        """Product of jacobian and matrix.
```
Comment (f-dangel) [inline test/core/derivatives/implementation/base.py:20]: Vectorized application of output-input Jacobian.

File: test/core/derivatives/implementation/base.py [hunk 21]
```diff
@@ -1,23 +1,163 @@
...
+
+        Args:
+            mat: matrix
```
Comment (f-dangel) [inline test/core/derivatives/implementation/base.py:23]: Collects vectors that will be multiplied by the Jacobian along its leading dimension.

File: test/core/derivatives/implementation/base.py [hunk 22]
```diff
@@ -1,23 +1,163 @@
...
+            product
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def jac_t_mat_prod(self, mat: Tensor) -> Tensor:
+        """Product of transposed jacobian and matrix.
+
+        Args:
+            mat: matrix
```
Comment (f-dangel) [inline test/core/derivatives/implementation/base.py:35]: Collects vectors that will be multiplied by the transpose Jacobian along its leading dimension.

File: test/core/derivatives/implementation/base.py [hunk 23]
```diff
@@ -1,23 +1,163 @@
...
+
+        Returns:
+            product
```
Comment (f-dangel) [inline test/core/derivatives/implementation/base.py:38]: Tensor representing the result of Jacobian-vector products. product[v] = mat[v] @ J.

File: test/core/derivatives/rnn_settings.py (hunk 1 above)
Comment (schaefertim) [reply test/core/derivatives/rnn_settings.py:28]: Yes. Removed.

File: test/core/derivatives/derivatives_test.py (hunk 14 above)
Comment (schaefertim) [reply test/core/derivatives/derivatives_test.py:75]: I'm not sure what you mean by this comment. In all cases ``AutogradDerivatives`` is used.

File: test/core/derivatives/derivatives_test.py (hunk 15 above)
Comment (schaefertim) [reply test/core/derivatives/derivatives_test.py:55]: This was an old comment. Deleted.

File: test/core/derivatives/implementation/backpack.py (hunk 17 above)
Comment (schaefertim) [reply test/core/derivatives/implementation/backpack.py:85]: At the moment, all the methods are documented in the abstract class ``DerivativesImplementation`` in base.py.
Thus, these methods are implemented in ``AutogradDerivatives``but don't need a new docstring.
Sadly, I did not find a better solution than just adding ``# noqa: D102``. In my case, the docstring is correctly copied from the parent class, but pydocstyle doesn't recognize it.

Do you still think a refactorization is the right way?

File: test/core/derivatives/implementation/base.py (hunk 20 above)
Comment (schaefertim) [reply test/core/derivatives/implementation/base.py:20]: I think "vectorized" is not correct for AutogradDerivatives.
I would change it to
Product of output-input-Jacobian and a matrix.

File: test/core/derivatives/derivatives_test.py (hunk 14 above)
Comment (schaefertim) [reply test/core/derivatives/derivatives_test.py:75]: Do you mean that this code should be moved inside the ``AutogradDerivatives(problem).jac_mat_prod(mat)`` function?

File: test/core/derivatives/implementation/backpack.py (hunk 17 above)
Comment (f-dangel) [reply test/core/derivatives/implementation/backpack.py:85]: For completeness, I found a similar question on the `pydocstyle` repo (https://github.com/PyCQA/pydocstyle/issues/274).

Comment (f-dangel) [reply test/core/derivatives/implementation/backpack.py:85]: Such a refactoring seems necessary for me in the future, but I understand that this PR is mainly concerned with the `RNN` partial derivative implementation. Can you open an issue on `backpack-discuss` to document this task for the future?

File: test/core/derivatives/derivatives_test.py (hunk 14 above)
Comment (schaefertim) [reply test/core/derivatives/derivatives_test.py:73]: Done.

File: test/core/derivatives/implementation/backpack.py (hunk 17 above)
Comment (schaefertim) [reply test/core/derivatives/implementation/backpack.py:85]: Created the issue: https://github.com/fKunstner/backpack-discuss/issues/101

**Comment by schaefertim** [comment]
This was an unlucky test run. Please restart.
Should I create an issue concerning the randomness of tests? The reason why I didn't create it thus far, is that I'm simply not sure which tests are still random.

**Comment by f-dangel** [comment]
> This was an unlucky test run. Please restart.
> Should I create an issue concerning the randomness of tests? The reason why I didn't create it thus far, is that I'm simply not sure which tests are still random.

There already exists a similar issue: https://github.com/fKunstner/backpack-discuss/issues/80. The old test suite is not fully deterministic.

File: backpack/core/derivatives/basederivatives.py [hunk 24]
```diff
@@ -289,73 +393,315 @@ def weight_jac_t_mat_prod(self, module, g_inp, g_out, mat, sum_batch=True):
...
 
         Return a function that maps mat to H * mat.
+
+        Args:
+            module: module to perform derivatives on
+            g_inp: input gradients
+            g_out: output gradients
+
+        Returns:
+            hessian matrix product
```
Comment (f-dangel) [inline backpack/core/derivatives/basederivatives.py:634]: Replace with line 626

File: test/core/derivatives/derivatives_test.py [hunk 25]
```diff
@@ -46,13 +47,19 @@
...
+
+@pytest.mark.parametrize(
+    "problem", NO_LOSS_PROBLEMS + RNN_PROBLEMS, ids=NO_LOSS_IDS + RNN_IDS
+)
+def test_jac_mat_prod(problem, request, V=3):
     """Test the Jacobian-matrix product.
 
     Args:
         problem (DerivativesProblem): Problem for derivative test.
+        request: pytest request of problem
```
Comment (f-dangel) [inline test/core/derivatives/derivatives_test.py:62]: Is this still required?

File: backpack/core/derivatives/rnn.py [hunk 26]
```diff
@@ -0,0 +1,236 @@
...
+        """
+        V: int = mat.shape[0]
+        N: int = mat.shape[2]
+        T: int = mat.shape[1]
+        H: int = mat.shape[3]
+        a_jac_t_mat_prod: Tensor = zeros(V, T, N, H, device=mat.device)
+        for t in range(T)[::-1]:
+            if t == (T - 1):
+                a_jac_t_mat_prod[:, t, ...] = einsum(
+                    "vnh, nh -> vnh",
```
Comment (f-dangel) [inline backpack/core/derivatives/rnn.py:73]: The code so far consistently uses no white space in `einsum` equations. Can you adapt all equations accordingly and remove the white space?

File: test/core/derivatives/derivatives_test.py (hunk 25 above)
Comment (schaefertim) [reply test/core/derivatives/derivatives_test.py:62]: No. Deleted.

File: backpack/core/derivatives/basederivatives.py (hunk 24 above)
Comment (schaefertim) [reply backpack/core/derivatives/basederivatives.py:634]: Then the type hints for return value were wrong as well.

**Comment by f-dangel** [comment]
Since this PR adds type annotations, can you also bump the `python` dependency [here](https://github.com/f-dangel/backpack/blob/rnn/setup.py)?

**Comment by schaefertim** [comment]
> Since this PR adds type annotations, can you also bump the `python` dependency [here](https://github.com/f-dangel/backpack/blob/rnn/setup.py)?

This still works with the old python version. Only circular dependencies introduced by type hints need a newer python version. Still not the case here. It's going to be obvious because will be a ``from future import``command.

**Pull request merged**